packages/core/helper-plugin/src/components/DynamicTable/v2/Table.js

```diff
@@ -0,0 +1,422 @@
+  headers: PropTypes.arrayOf(
+    PropTypes.shape({
+      cellFormatter: PropTypes.func,
+      key: PropTypes.string.isRequired,
+      metadatas: PropTypes.shape({
+        label: PropTypes.string.isRequired,
+        sortable: PropTypes.bool,
+      }).isRequired,
+      name: PropTypes.string.isRequired,
+    })
+  ),
+};
+
+/* -------------------------------------------------------------------------------------------------
+ * TableEmptyBody
+ * -----------------------------------------------------------------------------------------------*/
+
+const TableEmptyBody = ({ colSpan, isLoading, contentType, ...rest }) => {
+  const [{ query }] = useQueryParams();
+  const hasFilters = query?.filters !== undefined;
+
+  const content = hasFilters
+    ? {
+        id: 'content-manager.components.TableEmpty.withFilters',
+        defaultMessage: 'There are no {contentType} with the applied filters...',
+        values: { contentType },
+      }
+    : undefined;
+
+  if (isLoading) {
+    return (
+      <Tbody>
+        <Tr>
+          <Td colSpan={colSpan}>
+            <Flex justifyContent="center">
+              <Box padding={11} background="neutral0">
+                <Loader>Loading content...</Loader>
+              </Box>
+            </Flex>
+          </Td>
+        </Tr>
+      </Tbody>
+    );
+  }
+
+  return (
+    <Tbody>
+      <Tr>
+        <Td colSpan={colSpan}>
+          <EmptyStateLayout {...rest} content={content} hasRadius={false} shadow="" />
+        </Td>
+      </Tr>
+    </Tbody>
+  );
+};
+
+TableEmptyBody.defaultProps = {
+  action: undefined,
+  colSpan: 1,
+  content: undefined,
+  icon: undefined,
+  isLoading: false,
+};
+
+TableEmptyBody.propTypes = {
+  action: PropTypes.any,
+  colSpan: PropTypes.number,
+  content: PropTypes.shape({
+    id: PropTypes.string,
+    defaultMessage: PropTypes.string,
+    values: PropTypes.object,
+  }),
+  icon: PropTypes.oneOf(['document', 'media', 'permissions']),
+  isLoading: PropTypes.bool,
+  contentType: PropTypes.string.isRequired,
+};
+
+/* -------------------------------------------------------------------------------------------------
+ * Table
+ * -----------------------------------------------------------------------------------------------*/
+
+const Table = ({ children, footer, rowCount, colCount, tableActionBar, tableHead }) => {
+  const [selectedEntries, setSelectedEntries] = useState([]);
+
+  const handleSelectRow = ({ name, value }) => {
+    setSelectedEntries((prev) => {
+      if (value) {
+        return prev.concat(name);
+      }
+
+      return prev.filter((id) => id !== name);
+    });
+  };
+
+  return (
+    <TableProvider
+      selectedEntries={selectedEntries}
+      setSelectedEntries={setSelectedEntries}
+      onSelectRow={handleSelectRow}
+    >
+      {tableActionBar}
+      <TableCompo colCount={colCount} rowCount={rowCount} footer={footer}>
+        {tableHead}
+        {children}
+      </TableCompo>
+    </TableProvider>
+  );
```

Comment:
This pattern doesn't really work because it's still not extensible e.g. what happens if we need another action bar? If you look at the sort of thing done in the DS https://github.com/strapi/design-system/blob/95a02179f945511d2c5fb50efb880ed13f3169c9/packages/primitives/src/components/Combobox/Combobox.tsx#L215 we just pass `children` then it really doesn't matter what's passed.

Reply:
Sorry, i used that as an example not specifically saying just this one isn't ideal, the point is _ultimate flexibility_ take the `startAction/endAction` paradigm we have on things like a field input, you can add a `labelAction` to a field, but what if I wanted to add a small description below the label? I have to add a new prop. In short, whilst what you're doing now fits the brief really well, we can't possibly anticipate the needs in the future, which is how we're in this situation in a lot of places to being with imo.

Here's a blog post that probably explains better than I – https://epicreact.dev/soul-crushing-components/

Note, we did it with the information Box for CE/EE work and it was incredibly helpful – we did originally play with a more slot approach, @gu-stav can attest to this 👍🏼

Reply:
Gotcha I'm cool with abandoning the slot approach to avoid a headache in the future 👍. Thanks for sharing the insights and article. I had read this one https://epicreact.dev/one-react-mistake-thats-slowing-you-down/ where KCD demonstrates composition with props. Maybe it's older 🤷‍♂️